pkg/verify/verify.go

```diff
@@ -0,0 +1,469 @@
+}
+
+func (v *releaseVerifier) Verify(ctx context.Context, releaseDigest string) error {
+	if len(v.verifiers) == 0 || len(v.stores) == 0 {
+		return fmt.Errorf("the release verifier is incorrectly configured, unable to verify digests")
+	}
+	if len(releaseDigest) == 0 {
+		return fmt.Errorf("release images that are not accessed via digest cannot be verified")
+	}
+	if !validReleaseDigest.MatchString(releaseDigest) {
+		return fmt.Errorf("the provided release image digest contains prohibited characters")
+	}
+	parts := strings.SplitN(releaseDigest, ":", 3)
+	if len(parts) != 2 || len(parts[0]) == 0 || len(parts[1]) == 0 {
+		return fmt.Errorf("the provided release image digest must be of the form ALGO:HASH")
+	}
+	algo, hash := parts[0], parts[1]
+	name := fmt.Sprintf("%s=%s", algo, hash)
+
+	remaining := make(map[string]openpgp.EntityList, len(v.verifiers))
+	for k, v := range v.verifiers {
+		remaining[k] = v
+	}
+
+	verifier := func(path string, signature []byte) (bool, error) {
+		for k, keyring := range remaining {
+			content, _, err := verifySignatureWithKeyring(bytes.NewReader(signature), keyring)
+			if err != nil {
+				glog.V(4).Infof("keyring %q could not verify signature: %v", k, err)
+				continue
+			}
+			if err := verifyAtomicContainerSignature(content, releaseDigest); err != nil {
+				glog.V(4).Infof("signature %q is not valid: %v", path, err)
+				continue
+			}
+			delete(remaining, k)
+		}
+		return len(remaining) > 0, nil
+	}
+
+	for _, store := range v.stores {
+		if len(remaining) == 0 {
+			break
+		}
+		switch store.Scheme {
+		case "file":
+			dir := filepath.Join(store.Path, name)
+			if err := checkFileSignatures(ctx, dir, maxSignatureSearch, verifier); err != nil {
+				return err
+			}
+		case "http", "https":
+			copied := *store
+			copied.Path = path.Join(store.Path, name)
+			if err := checkHTTPSignatures(ctx, v.client, copied, maxSignatureSearch, verifier); err != nil {
+				return err
+			}
+		default:
+			return fmt.Errorf("internal error: the store %s is unrecognized", store)
+		}
+	}
+
+	if len(remaining) > 0 {
+		if glog.V(4) {
+			for k := range remaining {
+				glog.Infof("Unable to verify %s against keyring %s", releaseDigest, k)
+			}
+		}
+		return fmt.Errorf("unable to locate a valid signature for one or more sources")
+	}
+	return nil
+}
+
+// checkFileSignatures reads signatures as "signature-1", "signature-2", etc out of a directory until
+// either the provided fn returns an error, false, or no such file exists. No more than maxSignaturesToCheck
+// will be read.
+func checkFileSignatures(ctx context.Context, dir string, maxSignaturesToCheck int, fn func(path string, signature []byte) (bool, error)) error {
+	base := filepath.Join(dir, "signature-")
+	for i := 1; i < maxSignatureSearch; i++ {
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+
+		path := base + strconv.Itoa(i)
+		data, err := ioutil.ReadFile(path)
+		if err != nil {
+			if os.IsNotExist(err) {
+				if i == 1 {
```

Comment:
why this check?

Reply:
not truly necessary, I can remove it now that there are tests